### PR TITLE
Support `text-writing-mode` property for line placement symbols

### PR DESCRIPTION
**Pull request by zmiao**
## Launch Checklist

The line labels now can be configured with `text-writing-mode` as the following JSON shows:  
``` json
      {
        "id": "lines-symbol",
        "type": "symbol",
        "source": "mapbox",
        "layout": {
          "text-field": "新华路12号A栋",
          "symbol-placement": "line",
          "text-writing-mode":["vertical"]
        }
      }
```
The text-writing-mode configuration can be added as an array, whose values are enumeration values from a ( horizontal | vertical ) set.
To keep the previous rendering behavior as much as possible, the default `text-writing-mode` configuration for line label would be either:

`"text-writing-mode": ["horizontal", "vertical"] `
or
`"text-writing-mode": ["vertical", "horizontal"]`

It means for line labels, the order of the writing modes in the array doesn’t affect the rendering behavior if both writing modes are provided. The decision is based on the assumption that the default map text-rotation-alignment is applied for the line labels.

The rendering result will keep almost the same as previous except that for vertical line labels, non-CJK characters along with CJK characters will also be upright. The following table shows the rendering effects with different `text-writing-mode` configs, **to be aware that `vertical` writing mode will only be effective for text that is verticalizable, for pure Latin/numbers, the text will still be rendered horizontally with `vertical` writing mode**:
| 'text-writing-mode'       | Effects       |
| ------------- |:-------------:|
| ['horizontal']     | Screen Shot 2021-06-08 at 4 51 25 PM|
| ['vertical']      | Screen Shot 2021-06-08 at 4 51 34 PM|
| default/['horizontal', 'veritcal']/['vertical', 'horizontal']| Screen Shot 2021-06-08 at 4 51 43 PM |

 - [x] briefly describe the changes in this PR
 - [x] include before/after visuals or gifs if this PR includes visual changes
 - [x] write tests for all new functionality
 - [x] document any changes to public APIs
 - [x] post benchmark scores
 - [x] manually test the debug page
 - [x] tagged @mapbox/map-design-team` `@mapbox/static-apis if this PR includes style spec API or visual changes
 - [x] tagged @mapbox/gl-native if this PR includes shader changes or needs a native port
 - [x] apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'
 - [x] add an entry inside this element for inclusion in the `mapbox-gl-js` changelog: `<changelog>Support 'text-writing-mode' property for line symbol-placement text labels. Note: This change will bring following changes for CJK text block: 1. For vertical CJK text, all the characters including Latin and Numbers will be vertically placed now. Previously, Latin and Numbers are horizontally placed. 2. For horizontal CJK text, it may have a slight horizontal shift due to the anchor shift.</changelog>`

### CODE REVIEW

**Comment by zmiao** [comment]
Benchmark result:
Screen Shot 2021-06-09 at 2 55 28 PM
Screen Shot 2021-06-09 at 2 56 20 PM

File: src/symbol/placement.js [hunk 1]
```diff
@@ -956,17 +957,15 @@ export class Placement {
             if (hasIcon) {
                 const packedOpacity = packOpacity(opacityState.icon);
 
-                const useHorizontal = !(hasIconTextFit && symbolInstance.verticalPlacedIconSymbolIndex && horizontalHidden);
-
                 if (symbolInstance.placedIconSymbolIndex >= 0) {
-                    const horizontalOpacity = useHorizontal ? packedOpacity : PACKED_HIDDEN_OPACITY;
```
Comment (ansis) [inline src/symbol/placement.js:962]: This changed from `!hasIconTextFit || !verticalPlacedIconSymbolIndex || !horizontalHidden` to just `!horizontalHidden`. I understand why `!verticalPlacedIconSymbolIndex` would be dropped but I'm not sure about `!hasIconTextFit`. Do you know what the purpose of `!hasIconTextFit` was in the previous version?

File: src/symbol/projection.js [hunk 2]
```diff
@@ -175,6 +175,12 @@ function updateLineLabels(bucket: SymbolBucket,
     for (let s = 0; s < placedSymbols.length; s++) {
         const symbol: any = placedSymbols.get(s);
 
+        if (symbol.writingMode === WritingMode.vertical && !useVertical) {
+            if (s === 0 || placedSymbols.get(s - 1).writingMode !== WritingMode.horizontal) {
```
Comment (ansis) [inline src/symbol/projection.js:179]: Can you explain what this line is doing? especially the `s - 1` part

**Review by ansis — APPROVED** [review]
Looks great! I asked a couple questions about things that were unclear to me but I didn't see any issues

**Pull request merged**